pkg/storegateway/bucket_index_postings.go

```diff
@@ -322,3 +324,143 @@ func checkNilPosting(l labels.Label, p index.Postings) index.Postings {
+}
+
+func (s worstCaseFetchedDataStrategy) selectPostings(groups []postingGroup) (selected, omitted []postingGroup) {
+	const seriesBytesPerPostingByte = tsdb.EstimatedSeriesP99Size / tsdb.BytesPerPostingInAPostingList
+
+	sort.Slice(groups, func(i, j int) bool {
+		return groups[i].totalSize < groups[j].totalSize
+	})
+
+	var minGroupSize int64
+	for _, g := range groups {
+		// The size of each posting list contains 4 bytes with the number of entries.
+		// We shouldn't count these as series.
+		if !g.isSubtract && !(len(g.keys) == 1 && g.keys[0] == allPostingsKey) {
+			minGroupSize = g.totalSize - int64(len(g.keys)*4)
+			break
+		}
+	}
+
+	if minGroupSize == 0 {
+		// This should also cover the case of all postings group. all postings is requested only when there is no
+		// additive group.
+		return groups, nil
+	}
+
+	var (
+		selectedSize               int64
+		atLeastOneAdditiveSelected bool
+		maxSelectedSize            = minGroupSize * seriesBytesPerPostingByte
+	)
+	for i, g := range groups {
+		postingListSize := int64(float64(g.totalSize) * s.postingListActualSizeFactor)
+		if atLeastOneAdditiveSelected && selectedSize+postingListSize > maxSelectedSize {
+			return groups[:i], groups[i:]
+		}
+		selectedSize += postingListSize
+		atLeastOneAdditiveSelected = atLeastOneAdditiveSelected || !g.isSubtract
+	}
+	return groups, nil
+}
+
+// speculativeFetchedDataStrategy selects postings lists in a very similar way to worstCaseFetchedDataStrategy,
+// except it speculates on the size of the actual series after intersecting the selected posting lists.
+// Right now it assumes that each posting list will
+type speculativeFetchedDataStrategy struct{}
+
+func (s speculativeFetchedDataStrategy) name() string {
+	return "speculative"
+}
+
+func (s speculativeFetchedDataStrategy) selectPostings(groups []postingGroup) (selected, omitted []postingGroup) {
+	const (
+		postingsPerByteInPostingList = 4
+		bytesPerSeries               = 512
+
+		seriesBytesPerPostingByte = bytesPerSeries / postingsPerByteInPostingList
+	)
+
+	sort.Slice(groups, func(i, j int) bool {
+		return groups[i].totalSize < groups[j].totalSize
+	})
+
+	var minGroupSize int64
+	for _, g := range groups {
+		// The size of each posting list contains 4 bytes with the number of entries.
+		// We shouldn't count these as series.
+		if !g.isSubtract && !(len(g.keys) == 1 && g.keys[0] == allPostingsKey) {
+			minGroupSize = g.totalSize - int64(len(g.keys)*4)
+			break
+		}
+	}
+
+	if minGroupSize == 0 {
+		// This should also cover the case of all-postings group. All-postings is only included when there is no
+		// other intersecting group.
+		return groups, nil
+	}
+
+	var (
+		selectedSize                   int64
+		atLeastOneIntersectingSelected bool
+		maxSelectedSize                = minGroupSize * seriesBytesPerPostingByte
+	)
+	for i, g := range groups {
+		if atLeastOneIntersectingSelected && selectedSize+g.totalSize > maxSelectedSize {
```

Comment:
[nit] Here it's called `atLeastOneIntersectingSelected`, while in the other strategy it's called `atLeastOneAdditiveSelected`, but they're the same thing. We should uniform the naming.